service/subprocessmanager/manager.go

```diff
@@ -0,0 +1,146 @@
+	var argsSlice []string
+
+	// Parse the command line string into arguments
+	args, err := shellquote.Split(proc.Command)
+	if err != nil {
+		return 0, fmt.Errorf("could not parse command, error: %w", err)
+	}
+	// Separate the executable from the flags for the Command object
+	if len(args) > 1 {
+		argsSlice = args[1:]
+	}
+
+	// Create the command object and attach current os environment + environment variables defined by user
+	childProcess := exec.Command(args[0], argsSlice...)
+	childProcess.Env = append(os.Environ(), formatEnvSlice(&proc.Env)...)
+
+	// Handle the parent process being killed
+	handleParentProcessKill(childProcess)
+
+	// Handle the subprocess standard and error outputs in goroutines
+	stdoutReader, stdoutErr := childProcess.StdoutPipe()
+	if stdoutErr != nil {
+		return 0, fmt.Errorf("could not get the command's stdout pipe, err: %w", stdoutErr)
+	}
+	go proc.pipeSubprocessOutput(bufio.NewReader(stdoutReader), logger, true)
+
+	stderrReader, stderrErr := childProcess.StderrPipe()
+	if stderrErr != nil {
+		return 0, fmt.Errorf("could not get the command's stderr pipe, err: %w", stderrErr)
+	}
+	go proc.pipeSubprocessOutput(bufio.NewReader(stderrReader), logger, false)
+
+	// Start and stop timer (elapsed) right before and after executing the command
+	start := time.Now()
+	errProcess := childProcess.Start()
+	if errProcess != nil {
+		return 0, fmt.Errorf("process could not start: %w", errProcess)
+	}
+
+	errProcess = childProcess.Wait()
+	elapsed := time.Since(start)
+	if errProcess != nil {
+		return elapsed, fmt.Errorf("process error: %w", errProcess)
+	}
+
+	return elapsed, nil
+}
+
+// Log every line of the subprocesse's output using zap, until pipe is closed (EOF)
+func (proc *SubprocessSettings) pipeSubprocessOutput(reader *bufio.Reader, logger *zap.Logger, isStdout bool) {
+	for {
+		line, err := reader.ReadString('\n')
+		if err != nil && err != io.EOF {
+			logger.Info("subprocess logging failed", zap.String("error", err.Error()))
+			break
+		}
+
+		line = strings.TrimSpace(line)
+		if line != "" && line != "\n" {
+			if isStdout {
+				logger.Info("subprocess output line", zap.String("output", line))
+			} else {
+				logger.Error("subprocess output line", zap.String("output", line))
+			}
+		}
+
+		// Leave this function when error is EOF (stderr/stdout pipe was closed)
+		if err == io.EOF {
+			break
+		}
+	}
+}
+
+// formatEnvSlice will loop over the key-value pairs and format the slice correctly for use by the Command object ("name=value")
+func formatEnvSlice(envs *[]EnvSettings) []string {
+	if len(*envs) == 0 {
+		return nil
+	}
+
+	envSlice := make([]string, len(*envs))
+	for i, env := range *envs {
+		envSlice[i] = fmt.Sprintf("%v=%v", env.Name, env.Value)
+	}
+
+	return envSlice
+}
+
+// handleParentProcessKill kills the subprocess when the parent process dies, it is Linux-specific
+func handleParentProcessKill(cmd *exec.Cmd) {
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Pdeathsig: syscall.SIGTERM,
```

Comment:
I'm pretty sure this will prevent the collector from compiling on non-linux platforms.  At least it was the case when I messed with this stuff a while ago.  Does CircleCI try and build on other platforms?

Reply:
Thanks!

> > @keitwb can this code in the current form be used by the fluentbit extension to run the external subprocess?
> 
> @bogdandrutu not really. I need the ability to be able to gracefully terminate the process and also to be able to write to stdin. Why not just reuse what I did there with a few tweaks to make it more generic instead of writing a new one? Another thing to consider is if we ever need to restart the process on demand (e.g. because of config changes).

I think lots of logic will have to live in the different receiver/exporter/extensions that use any subprocess management implementation. The receiver I created also has needs that make the boundary of what this package can handle in terms of subprocess management very limited.

Reply:
Thanks, that's why the build is failing!

Reply:
It'd be good to find a way of handling this on other OSs as well (at least on Windows).

Reply:
@Nicolas-MacBeth also note the convention for files that use build flags is to postfix the file name with the OS, i.e. `manager_linux.go` & `manager_others.go` (Go actually uses the file name postfix to implicitly add build constraints but I always prefer to make them explicit as well: https://golang.org/pkg/go/build/#hdr-Build_Constraints)

Reply:
Thanks!